include/ensmallen_bits/cmaes/bipop_cmaes_impl.hpp

```diff
@@ -0,0 +1,113 @@
+{ /* Nothing to do. */ }
+
+template<typename CMAESType>
+BIPOPCMAES<CMAESType>::BIPOPCMAES(const size_t lambda,
+                                  const typename CMAESType::transformationPolicyType& transformationPolicy,
+                                  const size_t batchSize,
+                                  const size_t maxIterations,
+                                  const double tolerance,
+                                  const typename CMAESType::selectionPolicyType& selectionPolicy,
+                                  double stepSize,
+                                  const size_t maxRestarts) :
+    cmaes(lambda, transformationPolicy, batchSize, maxIterations, tolerance, selectionPolicy, stepSize),
+    maxRestarts(maxRestarts)
+{ /* Nothing to do. */  }
+
+template<typename CMAESType>
+template<typename SeparableFunctionType, typename MatType, typename... CallbackTypes>
+typename MatType::elem_type BIPOPCMAES<CMAESType>::Optimize(
+    SeparableFunctionType& function,
+    MatType& iterateIn,
+    CallbackTypes&&... callbacks)
+{
+    StoreBestCoordinates<MatType> sbc;
+    size_t totalFunctionEvaluations = 0;
+    size_t largePopulationBudget = 0;
+    size_t smallPopulationBudget = 0;
+
+    std::random_device rd;
+    std::mt19937 gen(rd());
+    std::uniform_real_distribution<> dis(0.0, 1.0);
+
+    // First single run with default population size
+    MatType iterate = iterateIn;
+    cmaes.Optimize(function, iterate, sbc, callbacks...);
+
+    size_t defaultLambda = cmaes.PopulationSize();
+    size_t currentLargeLambda = defaultLambda;
+
+    for (size_t i = 0; i < maxRestarts; ++i)
+    {
+        if (largePopulationBudget <= smallPopulationBudget || i == 0 || i == maxRestarts - 1)
+        {
+            // Large population regime
+            currentLargeLambda *= 2;
+            cmaes.PopulationSize() = currentLargeLambda;
+            cmaes.StepSize() = 2.0;
+
+            iterate = iterateIn;
+            cmaes.Optimize(function, iterate, sbc, callbacks...);
+            
+            size_t evaluations = cmaes.FunctionEvaluations();
+            totalFunctionEvaluations += evaluations;
+            largePopulationBudget += evaluations;
+        }
+        else
+        {
+            // Small population regime
+            double u = dis(gen);
+            size_t smallLambda = static_cast<size_t>(defaultLambda * std::pow(0.5 * currentLargeLambda / defaultLambda, u * u));
+            cmaes.PopulationSize() = smallLambda;
+            cmaes.StepSize() = 2 * std::pow(10, -2*dis(gen));
+
+            iterate = iterateIn;
+            cmaes.Optimize(function, iterate, sbc, callbacks...);
+            
+            size_t evaluations = cmaes.FunctionEvaluations();
+            totalFunctionEvaluations += evaluations;
+            smallPopulationBudget += evaluations;
+        }
+
+        // Check if the total number of evaluations has exceeded the limit
+        if (totalFunctionEvaluations >= 1e9) {
```

Comment:
Should this be configurable?

Reply:
Yeah, most likely. Plus there should be more parameters to control the termination criteria of the underlying CMA (from the other PR) which I will need to add. But since I still need to review the correctness of the termination criteria I don't want to add/delete stuff now.

Reply:
Okay, makes sense.